plugins/modules/firewall.py

```diff
@@ -0,0 +1,532 @@
+      },
+      "id":xxxxxx,
+      "updated":"xxxxxx"
+   }
+]
+'''
+
+try:
+    from linode_api4 import Firewall
+except ImportError:
+    # handled in module_utils.linode_common
+    pass
+
+linode_firewall_addresses_spec: dict = dict(
+    ipv4=dict(type='list', elements='str'),
+    ipv6=dict(type='list', elements='str')
+)
+
+linode_firewall_rule_spec: dict = dict(
+    label=dict(type='str', required=True),
+    action=dict(type='str', required=True),
+    addresses=dict(type='dict', options=linode_firewall_addresses_spec),
+    description=dict(type='str'),
+    ports=dict(type='str'),
+    protocol=dict(type='str')
+)
+
+linode_firewall_rules_spec: dict = dict(
+    inbound=dict(type='list', elements='dict', options=linode_firewall_rule_spec),
+    inbound_policy=dict(type='str'),
+    outbound=dict(type='list', elements='dict', options=linode_firewall_rule_spec),
+    outbound_policy=dict(type='str'),
+)
+
+linode_firewall_device_spec: dict = dict(
+    id=dict(type='int', required=True),
+    type=dict(type='str', default='linode')
+)
+
+linode_firewall_spec: dict = dict(
+    label=dict(type='str'),
+    devices=dict(type='list', elements='dict', options=linode_firewall_device_spec),
+    rules=dict(type='dict', options=linode_firewall_rules_spec),
+    status=dict(type='str')
+)
+
+# Fields that can be updated on an existing Firewall
+linode_firewall_mutable: list[str] = [
+    'status',
+    'tags'
+]
+
+
+class LinodeFirewall(LinodeModuleBase):
+    """Configuration class for a Linode firewall resource"""
+
+    def __init__(self):
+        self.module_arg_spec = linode_firewall_spec
+
+        self.results: dict = dict(
+            changed=False,
+            actions=[],
+            firewall=None,
+            devices=None
+        )
+
+        self._firewall: Firewall = None
+
+        super().__init__(module_arg_spec=self.module_arg_spec)
+
+    def get_firewall_by_label(self, label) -> Firewall:
+        """Gets a Linode Firewall by label"""
+
+        try:
+            return self.client.networking.firewalls(Firewall.label == label)[0]
+        except IndexError:
+            return None
+        except Exception as exception:
+            self.fail(msg='failed to get firewall {0}: {1}'.format(label, exception))
+
+    def create_firewall(self, spec_args: dict) -> dict:
+        """Creates a Linode Firewall"""
+
+        spec_args = copy.deepcopy(spec_args)
+
+        label = spec_args.pop('label')
+        spec_args = {k: v for k, v in spec_args.items() if k in {
+            'rules', 'tags'
+        }}
+
+        try:
+            result = self.client.networking.firewall_create(label, **spec_args)
+        except Exception as exception:
+            self.fail(msg='failed to create firewall: {0}'.format(exception))
+
+        return result
+
+    def __update_devices(self, spec_devices: list):
+        # Remove devices that are not present in config
+        remote_devices = self._firewall.devices
+
```

Comment:
This could be a lot simpler:

Accumulate device IDs in sets, organized by type. This might looks like

```
{'linode': {123, 124}}
```

then you can iterate through the spec like:
```
if device_id not in remote_devices[device_type]:
  create_device(device_id, device_type)
  remote_devices[device_type].remove(device_id)

for device_type in remote_devices:
  for device_id in remote_devices[device_type]:
    # this device was unclaimed...
    delete(device_type, device_id)
```